### ISSUE TITLE
Sax parser doesn't preserve same text whitespace as browser

### ISSUE THREAD

**Issue by tomjw64** [ohler55/ox#187]
It seems like sax parsing does not pick up on all newlines/indentation in some cases where it would matter in browser display. I've tried all whitespace skipping/preservation options available in `Ox.sax_parse`. In the example below, it seems the sax parser's output is identical for two files with different whitespace.

Update: Using `:skip => :skip_none`, explicitly, fixes the issue. It appears that the documentation listing :skip_none as the default value (i.e. here: http://www.rubydoc.info/github/ohler55/ox/Ox#sax_parse-class_method) is no longer up to date?

version: 2.6.0

```ruby
require 'ox'

class WhitespaceSax < Ox::Sax
  attr_reader :output
  def initialize
    @output = ''
  end
  def start_element(name)
    output << "Start: #{name}\n"
  end
  def end_element(name)
    output << "End: #{name}\n"
  end
  def text(str)
    output << "#{str.inspect}\n"
  end
end

## whitespace_one.html =>
#
# <html>
#   <body>
#     <span>span</span>
#     <span>span2</span>
#   </body>
# </html>
#
# Output in Chrome => "span span2 hello"

sax_parser_one = WhitespaceSax.new
File.open('whitespace_one.html', 'r') do |f|
  Ox.sax_parse(sax_parser_one, f)
end
File.open('whitespace_one.html', 'r') do |f|
  Ox.sax_parse(sax_parser_one, f, :skip => :skip_return)
end
File.open('whitespace_one.html', 'r') do |f|
  Ox.sax_parse(sax_parser_one, f, :skip => :skip_white)
end

## whitespace_two.html =>
#
# <html>
#   <body>
#     <span>span</span><span>span2</span>
#     hello
#   </body>
# </html>
#
# Output in Chrome => "spanspan2 hello"

sax_parser_two = WhitespaceSax.new
File.open('whitespace_two.html', 'r') do |f|
  Ox.sax_parse(sax_parser_two, f)
end
File.open('whitespace_two.html', 'r') do |f|
  Ox.sax_parse(sax_parser_two, f, :skip => :skip_return)
end
File.open('whitespace_two.html', 'r') do |f|
  Ox.sax_parse(sax_parser_two, f, :skip => :skip_white)
end

# Output of sax parsers equal?
puts sax_parser_one.output.eql? sax_parser_two.output # => true
```

**Comment by tomjw64** [comment]
Oh... Strangely, changing:
```ruby
sax_parser_one = WhitespaceSax.new
File.open('whitespace_one.html', 'r') do |f|
  Ox.sax_parse(sax_parser_one, f)
end
```
to
```ruby
sax_parser_one = WhitespaceSax.new
File.open('whitespace_one.html', 'r') do |f|
  Ox.sax_parse(sax_parser_one, f, :skip => :skip_none)
end
```
no longer results in an equal output. Whereas changing it to
```ruby
sax_parser_one = WhitespaceSax.new
File.open('whitespace_one.html', 'r') do |f|
  Ox.sax_parse(sax_parser_one, f, :skip => :skip_white)
end
```
continues to result in equal output.
Has the default option for `skip` changed from `:skip_none` to `:skip_white`?

**Comment by ohler55** [comment]
The default is `:skip_white` now. It changes in version 2.5.0 as noted in the CHANGELOG.md.

**Comment by ohler55** [comment]
I assume this can be closed. Right?

**Comment by tomjw64** [comment]
Yeah, I suppose so, since the actual problem was just some out of date RubyDoc documentation 😆

**Comment by tomjw64** [comment]
I am now having this issue even when I'm not being ignorant about whitespace skipping options.

version: 2.6.0, 2.7.0

```ruby
require 'ox'

class WhitespaceSax < Ox::Sax
  attr_reader :output
  def initialize
    @output = ''
  end
  # Output all available info just to cover all the bases
  def start_element(name); output << "Start: #{name}\n"; end
  def end_element(name); output << "End: #{name}\n"; end
  def text(str); output << "Text: #{str.inspect}\n"; end
  def instruct(target); output << "Instruct: #{target.to_s}\n"; end
  def end_instruct(target); output << "Instruct-End :#{target.to_s}\n"; end
  def attr(name,str); output << "Attr: #{name}, #{str}\n"; end
  def attrs_done(); output << "Attrs-Done\n"; end
  def doctype(str); ouptut << "Doctype: #{str}\n"; end
  def comment(str); output << "Comment: #{str}\n"; end
  def cdata(str); output << "CData: #{str}\n"; end
  def error(msg, line, column); ouptut << "Error: #{msg}, #{line}, #{column}\n"; end
  def abort(name); ouptut << "Abort: #{name}\n"; end
end

## whitespace_one.html =>
#
# <html>
#   <body>
#     <span>
#       <span>Hello</span>
#     </span>World
#   </body>
# </html>
#
# Output in Chrome => "Hello World"
sax_parser_one = WhitespaceSax.new
File.open('whitespace_one.html', 'r') do |f|
  Ox.sax_parse(sax_parser_one, f, :skip => :skip_none)
end

## whitespace_two.html =>
#
# <html>
#   <body>
#     <span>
#       <span>Hello</span></span>World
#   </body>
# </html>
#
# Output in Chrome => "HelloWorld"
sax_parser_two = WhitespaceSax.new
File.open('whitespace_two.html', 'r') do |f|
  Ox.sax_parse(sax_parser_two, f, :skip => :skip_none)
end

# Output of parsed files equal?
puts sax_parser_one.output.eql? sax_parser_two.output # => true
```

**Comment by ohler55** [comment]
I'll look into it tonight.

**Comment by ohler55** [comment]
I see, the whitespace after the span close is not included in either. I assume you are interested in that white space?

**Comment by tomjw64** [comment]
That's correct. I'm trying to make sure that I don't drop any whitespace in the files I'm working with (which often have elements with tail text, similar to the case above). Thanks for taking a look!

**Comment by ohler55** [comment]
If I want to avoid breaking other peoples code I think a new mode is required. The general consensus a long time back was that all white at the end of was not worthy of inclusion. Maybe a `:skip_absolutely_nothing` mode is needed. Maybe with a shorter name though.

**Comment by ohler55** [comment]
Let me make a branch with some debugging info in it. That might help identify what is going on.

**Comment by ohler55** [comment]
Just pushed a 'odd-chars' branch for testing purposes.

**Comment by tomjw64** [comment]
A new mode would be swell. Also, I can't seem to find the branch on the repo. Was that the intention?

**Comment by ohler55** [comment]
My mistake. That was actually an Oj debugging branch for a different issue. I have not made a branch for this issue yet. Since it sounds like a new mode will work I'll start a branch for that.

**Comment by ohler55** [comment]
Pushed a skip_off branch for you to try.

**Comment by tomjw64** [comment]
Works beautifully! I'm seeing the difference both in the test I wrote above and my other files. Thanks so much for taking your time to do this!

**Comment by ohler55** [comment]
Great. I'll get a release out today.

**Comment by ohler55** [comment]
Released 2.8.0.

**Comment by tomjw64** [comment]
Great! I'll go ahead and close the issue then.
